core/chains/evm/logpoller/log_poller.go

```diff
@@ -1191,17 +1244,300 @@ func (lp *logPoller) fillRemainingBlocksFromRPC(
+		cmp := bytes.Compare(newAddresses[j].Bytes(), addresses[i].Bytes())
+		if cmp < 0 {
+			merged = append(merged, newAddresses[j])
+			j++
+		} else if cmp > 0 {
+			merged = append(merged, addresses[i])
+			i++
+		} else {
+			merged = append(merged, addresses[i])
+			i++
+			j++ // only keep original, skip duplicate
+			continue
 		}
-		reqs = append(reqs, req)
 	}
 
+	// Append remaining elements, if any
+	merged = append(merged, newAddresses[j:]...)
+	merged = append(merged, addresses[i:]...)
+
+	req.SetAddresses(merged)
+}
+
+// isTopicsSubset returns true if all of the sets in the list topicsA are subsets of or equal to the sets in topicsB.
+// topicsA and topicsB each contain 4 (or any equal number of) sets of slices of topic values.
+//
+// Interpreting A & B as filters on the same contract address, "true" means that anything matching A will match B
+//
+// Assumptions:
+// - topicsA and topicsB must have the same length (outer dimension)
+// - every element of topicsA & topicsB are sorted lists containing no duplicates
+func isTopicsSubset(topicsA [][]common.Hash, topicsB [][]common.Hash) bool {
+	if len(topicsA) != len(topicsB) {
+		panic(fmt.Sprintf("Assumption violation: isTopicsSubset called with len(topicsA)=%d != len(topicsB)=%d", len(topicsA), len(topicsB)))
+	}
+
+	for i := range topicsA {
+		if len(topicsB[i]) == 0 {
+			continue // nil/empty list of topics matches all values, so topicsA[n] automatically a subset
+		}
+		if len(topicsA[i]) == 0 {
+			return false // topicsA[n] matches all values, but not topicsB[n], so topicsA is not a subset
+		}
+		topicsMapB := make(map[common.Hash]interface{})
+		for _, b := range topicsB[i] {
+			topicsMapB[b] = struct{}{}
+		}
+		for _, a := range topicsA[i] {
+			if _, ok := topicsMapB[a]; !ok {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func makeEventsTopicsKey(filter Filter) string {
+	// eventsTopicsKey is constructed to uniquely identify the particular combination of
+	// eventSigs and topics sets a filter has. Because we don't want the key to depend on
+	// the order of eventSigs, or the order of topic values for a specific topic index, we
+	// must make sure these 4 lists are sorted in the same way
+
+	size := len(filter.EventSigs[0])*(len(filter.EventSigs)+len(filter.Topic2)+len(filter.Topic3)+len(filter.Topic4)) + 4
+	var eventsTopicsKey = make([]byte, 0, size)
+
+	appendHashes := func(hashes []common.Hash) {
+		for _, h := range hashes {
+			eventsTopicsKey = append(eventsTopicsKey, h[:]...)
+		}
+		eventsTopicsKey = append(eventsTopicsKey, 0xFF) // separator
+	}
+	appendHashes(filter.EventSigs)
+	appendHashes(filter.Topic2)
+	appendHashes(filter.Topic3)
+	appendHashes(filter.Topic4)
+	return hex.EncodeToString(eventsTopicsKey)
+}
+
+func compareBlockNumbers(n, m *big.Int) (cmp int) {
+	if n != nil && m != nil {
+		return int(m.Uint64() - n.Uint64())
+	}
+	if n == nil {
+		cmp--
+	}
+	if m == nil {
+		cmp++
+	}
+	return cmp
+}
```

Comment:
It needs to return 0 when `n == nil && m == nil` and when `n != nil && m != nil`.  The switch statement you have would return -1 instead of 0. Could be fixed by adding another case... would be slightly more verbose but I guess still probably easier to read